pkg/runtime.v2/core/trainingruntime.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetv1alpha2 "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+	runtime "github.com/kubeflow/training-operator/pkg/runtime.v2"
+	fwkcore "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/core"
+	fwkplugins "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/plugins"
+	idxer "github.com/kubeflow/training-operator/pkg/runtime.v2/indexer"
+)
+
+var (
+	errorNotFoundSpecifiedTrainingRuntime = errors.New("not found TrainingRuntime specified in TrainJob")
+)
+
+type TrainingRuntime struct {
+	framework *fwkcore.Framework
+	client    client.Client
+	scheme    *apiruntime.Scheme
+}
+
+var TrainingRuntimeGroupKind = schema.GroupKind{
+	Group: kubeflowv2.GroupVersion.Group,
+	Kind:  "TrainingRuntime",
+}.String()
+
+var _ runtime.Runtime = (*TrainingRuntime)(nil)
+
+var trainingRuntimeFactory *TrainingRuntime
+
+func NewTrainingRuntime(ctx context.Context, c client.Client, indexer client.FieldIndexer) (runtime.Runtime, error) {
+	if err := indexer.IndexField(ctx, &kubeflowv2.TrainJob{}, idxer.TrainJobTrainingRuntimeRefKey, idxer.IndexTrainJobTrainingRuntimes); err != nil {
```

Comment:
This is a client informer indexer, not related to runtimeRef. This indexer allows us to get and list TrainJob by runtimeName. You can see the concrete example here: https://github.com/kubeflow/training-operator/blob/22da8af373f18e8e51c1e466a35a7738340b8c7d/pkg/runtime.v2/framework/plugins/coscheduling/coscheduling.go#L215

Reply:
@tenzen-y Do we really need to list TrainJob here that used by runtimes ? I guess, we can just list all TrainJobs, since TrainJob can't exists without runtime.

Reply:
Listing all trainJobs is redundant since here we need only TrainJobs with the specified runtime.
If we list all TrainJobs, after that we need to filter out unneeded trainJobs. That is redundant, right?

Reply:
The listing is used which TrainJob should be enqueue to reconciler queue.
Here, we do not enqueue all TrainJobs to reconciler queue.

Reply:
Oh, you are right. I am wondering in queueSuspendedTrainJob function, why we can't directly list all of the JobSets that have an appropriate Pod runtimeClass ?
Given that JobSet name is equal to TrainJob name, you can easily suspend the TrainJob.

Reply:
Maybe, you are missing direction here.
Here, we receive only RuntimeClass object from eventHandlers. 
We don't receive JobSet, TrainJob, and Runtimes.

Reply:
Yes, I see, maybe we should rename `queueSuspendedTrainJob` to `queueSuspendedTrainJobs` to make it more explicit, since you list all created TrainJobs  ?
Also, what name do we use for container runtime class ? For example, can 2 different runtime use the same container runtime class.

Reply:
> Yes, I see, maybe we should rename queueSuspendedTrainJob to queueSuspendedTrainJobs to make it more explicit, since you list all created TrainJobs ?

Sure. Indeed, the function will queue multiple TrainJobs.

> Also, what name do we use for container runtime class ? For example, can 2 different runtime use the same container runtime class.

The runtimeClass can be specified the container runtime name like containered, firecracker, and gvisor and so on. And then, each container runtime has overhead computing resources.

Reply:
Sounds good, thanks.

Reply:
Done.